Configurable name and namespace for avro schemas

SamirHafez opened the pull request:
### Description 
KSQL generated schemas always have a fixed name (`KsqlDataSourceSchema `) and namespace (`io.confluent.ksql.avro_schemas`).

A new config key is introduced (`AVRO_SCHEMA_FULL_NAME`) where we can set the full name to be used for a schema. This config key is per STREAM/TABLE created so there's flexibility beyond a general KSQL config variable. Usage should be something like:

```
CREATE STREAM pageviews_avro
  WITH (VALUE_FORMAT = 'AVRO'
              AVRO_SCHEMA_FULL_NAME = 'com.mycompany.PageViews') AS
  SELECT * FROM pageviews_json;
```

### Testing done 
A new test was added to make sure we use `AVRO_SCHEMA_FULL_NAME` if it exists in the config. 
- [ ] Add tests to `Analyzer`
- [ ] Add tests to `AvroDataTranslator`

### Reviewer checklist
- [x] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [x] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

SamirHafez commented:
[clabot:check]

SamirHafez commented:
@JimGalasyn Updated documentation, please take a look.

SamirHafez commented:
@apurvam turns out you were right.
Upon further testing I found out that the property gets consumed for `CREATE TABLE/STREAM` but not for `CREATE TABLE/STREAM AS` which runs through the Analyzer. I also needed to pipe the property there for it to work.

Please take another look and see if you agree.

JimGalasyn commented on docs/installation/server-config/avro-schema.rst:
```diff
@@ -38,6 +36,12 @@ Before using Avro in KSQL, make sure that |sr| is up and running and that ``ksql
+               (``io.confluent.ksql.avro_schemas``). You can override this behaviour by providing an ``AVRO_SCHEMA_FULL_NAME``
+               property in the ``WITH`` clause where you set the ``VALUE_FORMAT`` to ``'AVRO'``. As the name suggests,
+               this property will override the default name/namespace with the provided one.
+               E.g.: ``com.mycompany.MySchema`` registers a schema with the ``MySchema`` name and the ``com.mycompany`` namespace.
```
I'd replace "e.g." with "For example, ..."

SamirHafez commented:
@JimGalasyn All done.

big-andy-coates commented on ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroSerializerTest.java:
```diff
@@ -391,4 +392,38 @@ public void shouldTransformSourceNameDelimiterForInternal() {
   }
+
+    @Test
+    public void shouldUseSchemaNameFromPropertyIfExists() {
```
Can you add a test covering the default case too, i.e. where it uses the current schema name, if there isn't one already.

SamirHafez replied:
All the tests that use `shouldSerializeTypeCorrectly` (the tests for the different primitive types) already test that the name (and namespace) of the schema match `KsqlConstants.AVRO_SCHEMA_NAMESPACE` by default.

big-andy-coates commented on ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroSerializerTest.java:
```diff
@@ -391,4 +392,38 @@ public void shouldTransformSourceNameDelimiterForInternal() {
+        final String schemaName = "TestSchemaName1";
+        final String schemaNamespace = "com.test.namespace";
+
+        final Schema ksqlSchema = Schema.OPTIONAL_STRING_SCHEMA;
```
nit: inline

big-andy-coates commented on ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroSerializerTest.java:
```diff
@@ -391,4 +392,38 @@ public void shouldTransformSourceNameDelimiterForInternal() {
+        final String schemaNamespace = "com.test.namespace";
+
+        final Schema ksqlSchema = Schema.OPTIONAL_STRING_SCHEMA;
+        final Object ksqlValue = "foobar";
```
nit: inline.

big-andy-coates commented on ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroSerializerTest.java:
```diff
@@ -391,4 +392,38 @@ public void shouldTransformSourceNameDelimiterForInternal() {
+        final Schema ksqlSchema = Schema.OPTIONAL_STRING_SCHEMA;
+        final Object ksqlValue = "foobar";
+
+        final Map<String, String> properties = new HashMap<>();
```
nit: inline using `ImmutableMap.of(KsqlAvroTopicSerDe.AVRO_SCHEMA_FULL_NAME, schemaNamespace + "." + schemaName))`

big-andy-coates commented on ksql-serde/src/test/java/io/confluent/ksql/serde/avro/KsqlGenericRowAvroSerializerTest.java:
```diff
@@ -391,4 +392,38 @@ public void shouldTransformSourceNameDelimiterForInternal() {
+
+        final GenericRow ksqlRecord = new GenericRow(ImmutableList.of(ksqlValue));
+
+        final SchemaRegistryClient schemaRegistryClient = new MockSchemaRegistryClient();
```
nit: would you mind pulling this out into a field, (same for other tests if you have time).

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/RegisterTopicCommand.java:
```diff
@@ -53,16 +58,27 @@ public RegisterTopicCommand(final RegisterTopic registerTopic) {
     // if the property can be an unquoted identifier, then capitalization will have already happened
+    if (serde.toUpperCase() != DataSource.AVRO_SERDE_NAME
+        && properties.containsKey(KsqlAvroTopicSerDe.AVRO_SCHEMA_FULL_NAME)) {
+      log.warn("AVRO_SCHEMA_FULL_NAME is only valid for AVRO topics. Ignoring...");
```
Also, use `AVRO_SCHEMA_FULL_NAME + " is only valid..."`

SamirHafez commented:
@big-andy-coates thanks for the review.
I'll be absent for a little while while I deal with a few more pressing issues elsewhere. I will probably address your review at the start of next week if that's okay.

big-andy-coates commented:
No problem!

gnschenker commented:
@SamirHafez Is there any progress on this? I strongly support the idea of this PR since it is just bad to have no control over the `name` and `namespace` of the AVRO schema. The generic name used now does not at all reveal the intent of the schema. Downstream consumers might just be confused. 
Similar to @big-andy-coates I would also love to see the `namespace` configurable.

SamirHafez commented on ksql-serde/src/main/java/io/confluent/ksql/serde/avro/AvroDataTranslator.java:
```diff
@@ -38,11 +39,16 @@
 
+
   public AvroDataTranslator(final Schema ksqlSchema) {
+    this(ksqlSchema, new HashMap<String, String>());
```
I require further clarification here. Do you mean, remove the constructor and keep the one with both `ksqlSchema` and `properties` as arguments?

miguno commented:
Any updates here? We're happy to help move this along.

SamirHafez commented:
@miguno I've found some time to continue this. I think I addressed most comments by @big-andy-coates apart from the suggestion to make `io.confluent.ksql.avro_schemas` configurable. I suggest someone else addresses that once this is merged.

SamirHafez commented:
@big-andy-coates please give this PR another look

dmi3zkm commented:
Any plans to merge this PR ?

big-andy-coates commented:
Looking...

SamirHafez commented:
@big-andy-coates cool. I'll get to it. Hopefully I can keep the review addressing cycle shorter.

big-andy-coates commented:
Awesome!

SamirHafez commented:
Hey @big-andy-coates,
Added more tests and fixed the "inherited" avro schema situation. Please take another look when you get a chance.

big-andy-coates commented:
How you getting on @SamirHafez ?

BTW, once you've pushed your changes please re-add me as a reviewer so this PR pops back into my PR review list. 

Thanks

big-andy-coates commented:
Thanks @SamirHafez, I'm taking a look now.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java:
```diff
@@ -672,6 +693,7 @@ private void validateWithClause(final Set<String> withClauseVariables) {
     validSet.add(KsqlConstants.SINK_NUMBER_OF_PARTITIONS.toUpperCase());
     validSet.add(KsqlConstants.SINK_NUMBER_OF_REPLICAS.toUpperCase());
     validSet.add(DdlConfig.TIMESTAMP_FORMAT_PROPERTY.toUpperCase());
+    validSet.add(DdlConfig.AVRO_SCHEMA_FULL_NAME.toUpperCase());
```
Only valid if its a AVRO value format.  This function should throw if `AVRO_SCHEMA_FULL_NAME` is present, but the stream is not AVRO.

(Assuming you can determine if its AVRO at this point).

Ideally, add tests to cover both paths.

big-andy-coates replied:
Actually, I guess you're handling this else where - don't worry about it.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/ddl/commands/AbstractCreateStreamCommand.java:
```diff
@@ -196,6 +196,7 @@ private void validateWithClause(final Set<String> withClauseVariables) {
     validSet.add(DdlConfig.TOPIC_NAME_PROPERTY.toUpperCase());
     validSet.add(KsqlConstants.AVRO_SCHEMA_ID.toUpperCase());
     validSet.add(DdlConfig.TIMESTAMP_FORMAT_PROPERTY.toUpperCase());
+    validSet.add(DdlConfig.AVRO_SCHEMA_FULL_NAME.toUpperCase());
```
Only valid if its a AVRO value format.  This function should throw if `AVRO_SCHEMA_FULL_NAME` is present, but the stream is not AVRO.

(Assuming you can determine if its AVRO at this point).

Ideally, add tests to cover both paths.

big-andy-coates replied:
Actually, I guess you're handling this else where - don't worry about it.

big-andy-coates commented on ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java:
```diff
@@ -613,12 +619,20 @@ private void setIntoTopicName(final Table node) {
 
   private void setIntoTopicFormat(final StructuredDataSource into, final Table node) {
-    String serde = node.getProperties().get(DdlConfig.VALUE_FORMAT_PROPERTY).toString();
+    final Object serdeProperty = node.getProperties().get(DdlConfig.VALUE_FORMAT_PROPERTY);
```
Might be better structured as:

```java
final Object serdeProperty = node.getProperties().get(DdlConfig.VALUE_FORMAT_PROPERTY);

    String serde;
    if (serdeProperty != null) {
      serde = serdeProperty.toString();

      if (!serde.startsWith("'") && !serde.endsWith("'")) {
        throw new KsqlException(
            serde + " value is string and should be enclosed between " + "\"'\".");
      }
      serde = serde.substring(1, serde.length() - 1).toUpperCase();
    } else {
      final StructuredDataSource leftSource = analysis.getFromDataSource(0).left;
      serde =  leftSource.getKsqlTopic().getKsqlTopicSerDe().getSerDe().toString();
    }
```

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/analyzer/AnalyzerTest.java:
```diff
@@ -238,4 +253,223 @@ public void shouldCreateCorrectSinkKsqlTopic() {
+  }
+
+  @Test
+  public void shouldUseExplicitNamespaceWhenFormatIsInheritedForAvro() {
```
You can simplify this test by having:

```java
@Before
  public void init() {
    jsonMetaStore = MetaStoreFixture.getNewMetaStore(new InternalFunctionRegistry());
    avroMetaStore = MetaStoreFixture.getNewMetaStore(new InternalFunctionRegistry(), () -> new KsqlAvroTopicSerDe(KsqlConstants.DEFAULT_AVRO_SCHEMA_FULL_NAME));
  }

@Test
  public void shouldUseExplicitNamespaceWhenFormatIsInheritedForAvro() {
    final String simpleQuery = "create stream s1 with (AVRO_SCHEMA_FULL_NAME='org.ac.s1') as select * from test1;";

    final List<Statement> statements = getPreparedStatements(simpleQuery, avroMetaStore)
            .stream()
            .map(PreparedStatement::getStatement)
            .collect(Collectors.toList());
    final CreateStreamAsSelect createStreamAsSelect = (CreateStreamAsSelect) statements.get(0);
    final Table intoTable = new Table(QualifiedName.of(createStreamAsSelect.getName().toString()));
    intoTable.setProperties(createStreamAsSelect.getProperties());
    final QuerySpecification querySpecification = (QuerySpecification) createStreamAsSelect.getQuery().getQueryBody();
    final QuerySpecification newQuerySpecification = new QuerySpecification(
            querySpecification.getSelect(),
            intoTable,
            true,
            querySpecification.getFrom(),
            querySpecification.getWindowExpression(),
            querySpecification.getWhere(),
            querySpecification.getGroupBy(),
            querySpecification.getHaving(),
            querySpecification.getLimit()
    );
    final Analysis analysis = new Analysis();
    final Analyzer analyzer = new Analyzer("sqlExpression", analysis, avroMetaStore, "");
    analyzer.visitQuerySpecification(newQuerySpecification, new AnalysisContext(null));

    assertThat(analysis.getIntoProperties().get(DdlConfig.AVRO_SCHEMA_FULL_NAME).toString(), equalTo("'org.ac.s1'"));
  }
   
}
```

SamirHafez replied:
Ah yes. That makes sense.

big-andy-coates commented on ksql-engine/src/test/java/io/confluent/ksql/analyzer/AnalyzerTest.java:
```diff
@@ -238,4 +253,223 @@ public void shouldCreateCorrectSinkKsqlTopic() {
+  }
+
+  @Test
+  public void shouldUseImplicitNamespaceWhenFormatIsInheritedForAvro() {
```
as above.

SamirHafez commented:
@big-andy-coates pushed some more changes. Take a look when you're free.

big-andy-coates commented on docs/installation/server-config/avro-schema.rst:
```diff
@@ -37,6 +37,12 @@ Before using Avro in KSQL, make sure that |sr| is up and running and that ``ksql
 |cp|.
 
+.. important:: By default, KSQL registered avro schemas have the same name (``KsqlDataSourceSchema``) and the same namespace
+               (``io.confluent.ksql.avro_schemas``). You can override this behaviour by providing an ``AVRO_SCHEMA_FULL_NAME``
```
needs updating with the new name.

SamirHafez replied:
Updated

apurvam commented on ksql-engine/src/test/java/io/confluent/ksql/analyzer/AnalyzerTest.java:
```diff
@@ -238,4 +255,181 @@ public void shouldCreateCorrectSinkKsqlTopic() {
+  }
+
+  @Test
+  public void shouldUseExplicitNamespaceWhenFormatIsInheritedForAvro() {
```
What about a test around inheriting the namespace itself for `CREATE STREAM AS` and `CREATE TABLE AS` statements? Isn't this desirable behavior?

SamirHafez replied:
I would actually prefer to have the default always revert to `DEFAULT_AVRO_SCHEMA_FULL_NAME`.

Since the new property sets both the namespace and the name of the schema, I wouldn't like to see two different schemas with the same name (although seeing them with the same namespace could make sense).

Thoughts?

SamirHafez replied:
Wait, I see what you mean now. I'll add that test.

apurvam replied:
@SamirHafez one reason to inherit the namespace is that we already inherit the value type and also other properties (like number of partitions). I guess you are a target user, so what would you like to see? IF you are setting custom namespaces, would you want some outputs to default to the original?

apurvam replied:
cc @confluentinc/ksql , more opinions welcome.

dguy replied:
I wonder if we should be able to set the default namespace via a property? Then i think we could just stick with the way it is, i.e., use the default unless it is specified. Also, is it true that if we don't specify a `VALUE_AVRO_FULL_SCHEMA_NAME` that the schema name is identical for every query? If so that seems pretty bad ( I know that is not part of this PR). If it is then i think we should probably have a follow up PR that changes it to use the Table/Stream name or similar.

SamirHafez replied:
@dguy @big-andy-coates had suggested to allow a configurable default namespace back when I started this PR. We decided to keep that logic out of this PR and do it in a subsequent one.

SamirHafez replied:
@apurvam As a user I would explicitly set the namespace for my schemas, and if I didn't set it, I would expect it to be the default.

apurvam replied:
> Also, is it true that if we don't specify a VALUE_AVRO_FULL_SCHEMA_NAME that the schema name is identical for every query? 

It looks that way. I think enabling a user to set the default namespace is a good followup. However, that will still leave the possibility that all schemas have the same name. 

One of the risks I can see of changing the schema name inheritance behavior is upgrade compatibility: if we allow people to change the default schema namespace, then after an upgrade KSQL will probably try to register a second schema for the same topic, which probably will cause issues. 

I think for now, sticking with explicit naming makes sense. I'm happy with merging this if we can have a test case which ensures that we _don't_ inherit schema names and namespaces in CSAS / CTAS statements, @SamirHafez .

apurvam commented:
Thanks for the contribution @SamirHafez! I just merged this.
